src/components/Proposal/Create/SchemeForms/CreateGenericMultiCallProposal.tsx

```diff
@@ -0,0 +1,502 @@
+
+  initialFormValues: IFormValues;
+
+  constructor(props: IProps) {
+    super(props);
+
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.initialFormValues = importUrlValues<IFormValues>({
+      description: "",
+      title: "",
+      url: "",
+      tags: [],
+      addContract: "",
+      userContracts: [],
+      contracts: [
+        {
+          address: "",
+          value: 0,
+          abi: [],
+          methods: [],
+          method: "",
+          params: [],
+          values: [],
+          callData: "",
+        },
+      ],
+    });
+    this.state = {
+      loading: false,
+      tags: this.initialFormValues.tags,
+      addContractStatus: { error: "", message: "" },
+      whitelistedContracts: this.props.whitelistedContracts?.map(contract => { return contract.toLowerCase(); }) ?? [],
+      userContracts: [],
+    };
+  }
+
+  public async handleSubmit(formValues: IFormValues, { setSubmitting }: any): Promise<void> {
+    if (!await enableWalletProvider({ showNotification: this.props.showNotification })) { return; }
+
+    const contractsToCall = [];
+    const callsData = [];
+    const values = [];
+
+    for (const contract of formValues.contracts) {
+      contractsToCall.push(contract.address);
+      callsData.push(contract.callData);
+      values.push(contract.value);
+    }
+
+    const proposalValues = {
+      title: formValues.title,
+      description: formValues.description,
+      contractsToCall: contractsToCall,
+      callsData: callsData,
+      values: values,
+      url: formValues.url,
+      dao: this.props.daoAvatarAddress,
+      scheme: this.props.scheme.address,
+      tags: this.state.tags,
+    };
+
+    setSubmitting(false);
+    await this.props.createProposal(proposalValues);
+
+    Analytics.track("Submit Proposal", {
+      "DAO Address": this.props.daoAvatarAddress,
+      "Proposal Title": formValues.title,
+      "Scheme Address": this.props.scheme.address,
+      "Scheme Name": this.props.scheme.name,
+    });
+
+    this.props.handleClose();
+  }
+
+  // Exports data from form to a shareable url.
+  public exportFormValues(values: IFormValues) {
+    exportUrl({ ...values, ...this.state });
+    this.props.showNotification(NotificationStatus.Success, "Exportable url is now in clipboard :)");
+  }
+
+  private onTagsChange = (tags: any[]): void => {
+    this.setState({ tags });
+  }
+
+  /**
+ * Given a contract address, checks whether it's valid, not exists in the current contract list and that the contract is verified with valid ABI data and write methods.
+ * If all checks are okay, pushes the contract address to the contract lists, otherwise returns an appropriate message.
+ * @param {string} contractToCall
+ */
+  private verifyContract = async (contractToCall: string) => {
+    const addContractStatus: IAddContractStatus = {
+      error: "",
+      message: "",
+    };
+
+    if (contractToCall === "") {
+      addContractStatus.error = "";
```

Comment:
in that case no need error msg?

Reply:
No, this field is not really part of the form, it's just to add a new contract to the select picker. It's okay if it's empty.

Reply:
so why checking that at all ?

Reply:
When the user tries to add a new custom contract we need to check:

- Valid address
- The contract not already exist in the list (that he does not add the same contract twice)
- That the contract has a valid ABI data (verified with write methods)

If one of the above conditions does not met, there is not reason to add this contract to the select picker.

Reply:
Ahh because if an error already occurred and the user cleans the input so we want to remove the last error.